File: backend/models/hike.go
package models

import (
	"time"
)

type Hike struct {
	ID            uint           `gorm:"primaryKey" json:"id" example:"1"`
	CreatedAt     time.Time      `json:"created_at" example:"2024-05-28T12:54:10.517438+02:00"`
	UpdatedAt     time.Time      `json:"updated_at" example:"2024-05-28T12:54:10.517438+02:00"`
	Name          string         `gorm:"not null" json:"name" example:"Montagne du destin" validate:"required,max=100"`
	Description   string         `json:"description" example:"La rando de zinzin" validate:"required,max=500"`
	OrganizerID   uint           `json:"organizer_id" example:"1" validate:"required" gorm:"constraint:OnUpdate:CASCADE,OnDelete:CASCADE"`
	Difficulty    string         `json:"difficulty" example:"Moderate" validate:"required,oneof=Easy Moderate Hard"`
	Duration      int            `json:"duration" example:"3" validate:"required"`
	IsApproved    bool           `json:"is_approved" default:"false" example:"false"`
	Groups        []Group        `json:"groups"`
	Image         string         `json:"image" example:"hike_image.jpg"`
	GpxFile       string         `json:"gpx_file" example:"hike.gpx"`
	Subcriptions  []Subscription `json:"subscriptions"`
	AverageRating float64        `json:"average_rating" gorm:"-"`
	Lat           string         `json:"lat" example:"45.123456"`
	Lng           string         `json:"lng" example:"5.123456"`
}

File: frontend/lib/shared/models/hike.dart
import 'package:frontend/shared/models/subscriptions.dart';

class Hike {
  final int id;
  final String name;
  final String description;
  final String difficulty;
  final int duration;
  final bool isApproved;
  final String image;
  final String gpxFile;
  final List<Subscriptions> subscriptions;
  final double averageRating;
  final String lat;
  final String lng;


  Hike({
    required this.id,
    required this.name,
    required this.description,
    required this.difficulty,
    required this.duration,
    required this.isApproved,
    required this.image,
    required this.gpxFile,
    required this.subscriptions,
    required this.averageRating,
    required this.lat,
    required this.lng,
  });

  factory Hike.fromJson(Map<String, dynamic> json) {
    return Hike(
        id: json['id'],
        name: json['name'],
        description: json['description'],
        difficulty: json['difficulty'],
        duration: json['duration'],
        isApproved: json['is_approved'],
        image: json['image'],
        gpxFile: json['gpx_file'],
        subscriptions: json['subscriptions'] != null
            ? List<Subscriptions>.from(json['subscriptions']
                .map((user) => Subscriptions.fromJson(user)))
            : [],
        averageRating: (json['average_rating'] is int)
            ? (json['average_rating'] as int).toDouble()
            : (json['average_rating'] ?? 0.0)
          , lat: json['lat']
          , lng: json['lng']
            );
  }
}
